File: samples/java/DeliveryDispatch/Server/CourierSpotNode/src/main/java/systems/zlink/samples/deliverydispatch/server/courierspotnode/spots/handlers/CourierDecisionActorHandler.java
package systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.channels.ZLinkClient;
import systems.zlink.framework.spots.ZLinkEntrySpotActorSendHandler;
import systems.zlink.samples.deliverydispatch.server.configuration.SampleNames;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.CourierActor;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

/**
 * The courier's decision, sent back to dispatch one-way. The node neither judges it nor times it —
 * it carries the attempt the offer was made under, and dispatch decides whether that attempt is
 * still the current one (common sample spec section 7.4).
 */
public final class CourierDecisionActorHandler
        implements ZLinkEntrySpotActorSendHandler<
                CourierEntrySpot, CourierActor, Messages.CourierDecisionMsg> {
    private final ZLinkClient channels;

    public CourierDecisionActorHandler(ZLinkClient channels) {
        this.channels = channels;
    }

    @Override
    public CompletionStage<Void> handle(
            CourierEntrySpot entrySpot,
            CourierActor actor,
            ZLinkMessageContext context,
            Messages.CourierDecisionMsg message) {
        // --8<-- [start:doc-dd-decision-send]
        Optional<Integer> attempt = actor.takeOfferedAttempt(message.deliveryId());
        if (attempt.isEmpty()) {
            System.err.println(
                    "deliverydispatch courier-actor: decision for an unknown offer"
                            + " delivery="
                            + message.deliveryId()
                            + " courier="
                            + actor.actorId());
            return CompletableFuture.completedFuture(null);
        }

        return channels.sendToChannel(
                        SampleNames.DispatchChannel,
                        new Messages.OfferDeliveryResultMsg(
                                message.deliveryId(),
                                message.courierId(),
                                attempt.get(),
                                message.accepted(),
                                message.reason()))
                .submit()
                .thenRun(
                        () ->
                                System.out.println(
                                        "deliverydispatch courier-actor: decision delivery="
                                                + message.deliveryId()
                                                + " courier="
                                                + actor.actorId()
                                                + " attempt="
                                                + attempt.get()
                                                + " accepted="
                                                + message.accepted()));
        // --8<-- [end:doc-dd-decision-send]
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSpotNode/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/courierspotnode/spots/handlers/CourierDecisionActorHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.channels.ZLinkClient
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorSendHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.configuration.SampleNames
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.CourierActor
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.CourierDecisionMsg
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.OfferDeliveryResultMsg

/**
 * The courier's decision, sent back to dispatch one-way. The node neither judges it nor times it —
 * it carries the attempt the offer was made under, and dispatch decides whether that attempt is
 * still the current one (common sample spec section 7.4).
 */
class CourierDecisionActorHandler(private val channels: ZLinkClient) :
    ZLinkSuspendingEntrySpotActorSendHandler<CourierEntrySpot, CourierActor, CourierDecisionMsg> {
    override suspend fun handle(
        entrySpot: CourierEntrySpot,
        actor: CourierActor,
        context: ZLinkMessageContext,
        message: CourierDecisionMsg,
    ) {
        // --8<-- [start:doc-dd-decision-send]
        val attempt = actor.takeOfferedAttempt(message.deliveryId)
        if (attempt == null) {
            System.err.println(
                "deliverydispatch courier-actor: decision for an unknown offer " +
                    "delivery=${message.deliveryId} courier=${actor.actorId()}"
            )
            return
        }

        channels
            .sendToChannel(
                SampleNames.DispatchChannel,
                OfferDeliveryResultMsg(
                    deliveryId = message.deliveryId,
                    courierId = message.courierId,
                    attempt = attempt,
                    accepted = message.accepted,
                    reason = message.reason,
                ),
            )
            .submit()
        // --8<-- [end:doc-dd-decision-send]
        println(
            "deliverydispatch courier-actor: decision delivery=${message.deliveryId} " +
                "courier=${actor.actorId()} attempt=$attempt accepted=${message.accepted}"
        )
    }
}
